consensus/snapshot/snapshots_service.c

```diff
@@ -0,0 +1,420 @@
+  retcode_t ret;
+
+  iota_snapshot_read_lock(&snapshots_service->snapshots_provider->inital_snapshot);
+
+  if (target_milestone->index <= snapshots_service->snapshots_provider->inital_snapshot.metadata.index) {
+    log_error(logger_id, "Target milestone is too old\n");
+    ret = RC_SNAPSHOT_SERVICE_MILESTONE_TOO_OLD;
+    goto cleanup;
+  }
+  ERR_BIND_GOTO(iota_snapshot_copy(&snapshots_service->snapshots_provider->inital_snapshot, snapshot), ret, cleanup);
+  ERR_BIND_GOTO(iota_milestone_service_replay_milestones(snapshots_service->milestone_service, tangle, snapshot,
+                                                         target_milestone->index),
+                ret, cleanup);
+
+cleanup:
+  iota_snapshot_unlock(&snapshots_service->snapshots_provider->inital_snapshot);
+
+  return ret;
+}
+
+retcode_t iota_snapshots_service_generate_snapshot_metadata(snapshots_service_t *const snapshots_service,
+                                                            iota_milestone_t const *const target_milestone,
+                                                            tangle_t const *const tangle, snapshot_t *const snapshot) {
+  retcode_t ret;
+  DECLARE_PACK_SINGLE_TX(tx, tx_ptr, tx_pack);
+  snapshot->metadata.index = target_milestone->index;
+  memcpy(snapshot->metadata.hash, target_milestone->hash, FLEX_TRIT_SIZE_243);
+  ERR_BIND_RETURN(iota_tangle_transaction_load(tangle, TRANSACTION_FIELD_HASH, target_milestone->hash, &tx_pack), ret);
+  snapshot->metadata.timestamp = transaction_timestamp(&tx);
+  if ((snapshots_service->snapshots_provider->latest_snapshot.metadata.index - target_milestone->index) <=
+      SNAPSHOT_SERVICE_SOLID_ENTRY_POINT_MAX_DEPTH) {
+    ERR_BIND_RETURN(
+        iota_snapshots_service_update_solid_entry_points(snapshots_service, snapshot, target_milestone, tangle), ret);
+  }
+
+  // TODO - seen milestones
+  return RC_OK;
+}
+
+retcode_t iota_snapshots_service_persist_snapshot(snapshots_service_t *const snapshots_service,
+                                                  snapshot_t *const snapshot) {
+  retcode_t ret;
+
+  log_info(logger_id, "Persisting snapshot with index: %" PRIu64 "\n", snapshot->metadata.index);
+  ERR_BIND_GOTO(iota_snapshots_provider_write_snapshot_to_file(
+                    snapshot, snapshots_service->conf->local_snapshots.local_snapshots_path_base),
+                ret, cleanup);
+
+  log_info(logger_id, "Local snapshot files were succesfully written\n");
+  iota_snapshot_write_lock(&snapshots_service->snapshots_provider->inital_snapshot);
+
+  ERR_BIND_GOTO(iota_snapshot_copy(snapshot, &snapshots_service->snapshots_provider->inital_snapshot), ret, cleanup);
+  log_info(logger_id, "Local snapshot updated with snapshot index %" PRIu64 "\n",
+           snapshots_service->snapshots_provider->inital_snapshot.metadata.index);
+
+cleanup:
+  iota_snapshot_unlock(&snapshots_service->snapshots_provider->inital_snapshot);
```

Comment:
Don't unlock in the cleanup, it can cause undefined behaviour if `iota_snapshots_provider_write_snapshot_to_file` fails